### PR TITLE
ARROW-6821: [C++][Parquet] Do not require Thrift compiler when building (but still require library)

### PR DESCRIPTION
**Pull request by nealrichardson**
I tried to follow #5464 for inspiration since it did similar for flatbuffers (though this doesn't vendor Thrift).

### CODE REVIEW

**Comment by nealrichardson** [comment]
Just checking on the R build that does a bundled static build and was also building m4, flex, and bison in order to build thrift, the run time is down from 12 minutes on master to 8 minutes here. That's probably the biggest speedup we'll see since it was doing the most extra work, but any other builds that are bundling thrift should also see some speedups.

**Comment by wesm** [comment]
I'll help with this when I can. Biggest question I have is whether the generated files will be robust to Thrift version changes (will help with determining this)

**Comment by nealrichardson** [comment]
I've confirmed at least that the files generated with thrift 0.13 (latest release) work with 0.12 (the bundled version). 

Not sure what's up with Appveyor: it passes on my fork: https://ci.appveyor.com/project/nealrichardson/arrow/builds/31315270 Is there maybe stale build cache that needs to be cleared?

File: cpp/src/parquet/parquet_constants.h
Comment (pitrou) [inline cpp/src/parquet/parquet_constants.h]: Put those files in a `thrift_generated` subdirectory perhaps?

Comment (nealrichardson) [reply cpp/src/parquet/parquet_constants.h]: Would that fix the build issue, or you mean just to be explicit that they're generated?

Comment (pitrou) [reply cpp/src/parquet/parquet_constants.h]: Just be explicit that they're generated.

**Comment by nealrichardson** [comment]
Here's a log snippet from the Appveyor failure:

```
54/246] Performing build step for 'snappy_ep'
-- snappy_ep build command succeeded.  See also C:/projects/arrow/cpp/build/snappy_ep-prefix/src/snappy_ep-stamp/snappy_ep-build-*.log
[55/246] Creating directories for 'mimalloc_ep'
[56/246] Performing install step for 'snappy_ep'
-- snappy_ep install command succeeded.  See also C:/projects/arrow/cpp/build/snappy_ep-prefix/src/snappy_ep-stamp/snappy_ep-install-*.log
[57/246] Completed 'snappy_ep'
[58/246] Completed 'googletest_ep'
[59/246] Building CXX object src\parquet\CMakeFiles\parquet_test_support.dir\Unity\unity_0.cxx.obj
FAILED: src/parquet/CMakeFiles/parquet_test_support.dir/Unity/unity_0.cxx.obj 
C:\Miniconda37-x64\Scripts\clcache.exe  /nologo /TP -DARROW_MIMALLOC -DARROW_USE_SIMD -DARROW_WITH_BROTLI -DARROW_WITH_LZ4 -DARROW_WITH_SNAPPY -DARROW_WITH_ZLIB -DARROW_WITH_ZSTD -DBOOST_ALL_NO_LIB -DGTEST_LINKED_AS_SHARED_LIBRARY=1 -DHAVE_INTTYPES_H -DHAVE_NETDB_H -DNOMINMAX -DPARQUET_EXPORTING -D_CRT_SECURE_NO_WARNINGS -Isrc -I..\src -I..\src\generated -I..\thirdparty\flatbuffers\include -IC:\Miniconda37-x64\envs\arrow\Library\include -Isnappy_ep\src\snappy_ep-install\include -Ibrotli_ep\src\brotli_ep-install\include -Igflags_ep-prefix\src\gflags_ep\include -Ithrift_ep\src\thrift_ep-install\include -Imimalloc_ep\src\mimalloc_ep\lib\mimalloc-1.0\include -Igoogletest_ep-prefix\src\googletest_ep\include -Irapidjson_ep\src\rapidjson_ep-install\include -Ilz4_ep-prefix\include -Izstd_ep-install\include -I..\thirdparty\hadoop\include /DWIN32 /D_WINDOWS  /GR /EHsc /D_SILENCE_TR1_NAMESPACE_DEPRECATION_WARNING   /EHsc /bigobj /W3 /wd4800 /wd4996 /wd4065 /WX /MP /MD /Od /UNDEBUG /showIncludes /Fosrc\parquet\CMakeFiles\parquet_test_support.dir\Unity\unity_0.cxx.obj /Fdsrc\parquet\CMakeFiles\parquet_test_support.dir\parquet_test_support.pdb /FS -c src\parquet\CMakeFiles\parquet_test_support.dir\Unity\unity_0.cxx
c:\projects\arrow\cpp\src\parquet\parquet_types.h(12): fatal error C1083: Cannot open include file: 'thrift/Thrift.h': No such file or directory
[60/246] Performing download step (download, verify and extract) for 'mimalloc_ep'
-- Downloading...
   dst='C:/projects/arrow/cpp/build/mimalloc_ep-prefix/src/270e765454f98e8bab9d42609b153425f749fff6.tar.gz'
   timeout='none'
-- Using src='https://github.com/microsoft/mimalloc/archive/270e765454f98e8bab9d42609b153425f749fff6.tar.gz'
-- [download 100% complete]
-- Downloading... done
-- extracting...
     src='C:/projects/arrow/cpp/build/mimalloc_ep-prefix/src/270e765454f98e8bab9d42609b153425f749fff6.tar.gz'
     dst='C:/projects/arrow/cpp/build/mimalloc_ep-prefix/src/mimalloc_ep'
-- extracting... [tar xfz]
-- extracting... [analysis]
-- extracting... [rename]
-- extracting... [clean up]
-- extracting... done
[61/246] Performing download step (download, verify and extract) for 'thrift_ep'
-- thrift_ep download command succeeded.  See also C:/projects/arrow/cpp/build/thrift_ep-prefix/src/thrift_ep-stamp/thrift_ep-download-*.log
ninja: build stopped: subcommand failed.
```

`Building CXX object src\parquet\CMakeFiles\parquet_test_support.dir\Unity\unity_0.cxx.obj` needs to wait for `thrift_ep` but it isn't.

**Comment by pitrou** [comment]
Rebased and pushed a potential fix.

File: cpp/src/parquet/parquet_constants.h
Comment (wesm) [reply cpp/src/parquet/parquet_constants.h]: Since we already have src/generated that seems like an OK place

**Comment by nealrichardson** [comment]
Ok, I tested locally using older versions of thrift, and I found that the code checked in here (compiled with thrift 0.13, current release, October 2019) works with 0.12 (January 2019, the version we have pinned in versions.txt) and 0.11 (December 2017) but it does not with 0.10 (January 2017). [0.10 is the minimum we require currently](https://github.com/apache/arrow/blob/6fa6c91/cpp/cmake_modules/ThirdpartyToolchain.cmake#L1157).

Our options include, in increasing order of effort/maintenance burden required: 

1. Bump the minimum required thrift version from 0.10 to 0.11
2. Try generating the code with an older thrift compiler (0.11 or 0.12) and see if that gives us compatibility to the present as well as 0.10.
3. In parquet's cmake, if thrift version is detected as 0.10, (re)compile the thrift (i.e. do what we previously did every time, just conditional on thrift version).

**Comment by pitrou** [comment]
I would say either option 2 or 1. Definitely not option 3.

**Comment by nealrichardson** [comment]
Compiling with thrift 0.11 still does not seem to work on 0.10, so my preference is for bumping the minimum required thrift version from 0.10 to 0.11.

**Comment by nealrichardson** [comment]
@pitrou sadly the appveyor failure looks the same as before your change

**Comment by wesm** [comment]
@kou do you know if increasing the required Thrift version from 0.10 to 0.11 will affect any of the Linux package builds? That's the only place I would be concerned. In general I'm in favor of increasing the minimum required version

**Comment by pitrou** [comment]
@nealrichardson I think I have a better fix, trying :-)

**Comment by nealrichardson** [comment]
Most of our docker files have Thrift_SOURCE=BUNDLED already

**Comment by wesm** [comment]
I did a couple other small things (fixing Doxygen complaining about these files), and moved the generated files to src/generated

File: cpp/apidoc/Doxyfile
Comment (nealrichardson) [inline cpp/apidoc/Doxyfile]:
```suggestion
# (including Cygwin) and Mac users are advised to set this option to NO.
```

File: cpp/apidoc/Doxyfile
Comment (nealrichardson) [inline cpp/apidoc/Doxyfile]: @wesm Is this file autogenerated? I'm guessing we either need to add back the ASF header or add this to the 🐀 exclude file.

Comment (nealrichardson) [reply cpp/apidoc/Doxyfile]: Indeed: https://github.com/apache/arrow/pull/6557/checks?check_run_id=496256402#step:5:91

**Comment by nealrichardson** [comment]
I'll note again that these various fix candidates are passing appveyor on my fork: https://ci.appveyor.com/project/nealrichardson/arrow/builds/31352578

Is there build cache that needs to be cleared somehow?

**Comment by pitrou** [comment]
No, it needs to be fixed properly, as attempted.

**Comment by nealrichardson** [comment]
Right--your fixes are passing on my fork: https://ci.appveyor.com/project/nealrichardson/arrow/builds/31345208

**Comment by pitrou** [comment]
I know, but somehow the timing is different on the ASF AppVeyor. The different timing shouldn't be triggering a failure, so we must make sure it passes there too.

**Comment by nealrichardson** [comment]
Ah right, it's a race.

**Comment by kou** [comment]
> do you know if increasing the required Thrift version from 0.10 to 0.11 will affect any of the Linux package builds?

It doesn't affect the Linux package builds.
We don't need to care about the Linux package builds for this.

**Comment by nealrichardson** [comment]
While we're at it, should I bump the bundled thrift version to 0.13.0, the latest release? No reason I can see (other than inertia) that it should be 0.12.0.

**Comment by wesm** [comment]
@nealrichardson let's do the bundled version bump in a separate PR

**Comment by wesm** [comment]
Appveyor build in progress here https://ci.appveyor.com/project/nealrichardson/arrow/builds/31354803

I'll merge this if that passes. I'd guess some nightlies may break on account of this but better that we know sooner rather than later

**Comment by nealrichardson** [comment]
We've been passing on my fork for a while, not sure why ASF's Appveyor isn't passing. Since it's a race, Antoine's probably right (he usually is ;) that we should nail it down here, unless there's some other good explanation for why one would pass and the other fail.

**Comment by wesm** [comment]
I see. I opened https://issues.apache.org/jira/browse/ARROW-8049 for the version upgrade

**Comment by nealrichardson** [comment]
https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/31354809 is next in the queue so we should get the verdict in about an hour or so, sooner if it fails like the previous attempts.

**Comment by nealrichardson** [comment]
Alright, this appveyor build has gotten past where the others have failed so I'm optimistic. 

Before merging, should I turn back off the VERBOSE_THIRDPARTY_BUILD flags? Any other cleanup from the thrashing around trying to fix the issue?

**Comment by wesm** [comment]
Ah yes do disable the verbose thirdparty build. I didn't see anything else concerning in a quick glance through

**Comment by nealrichardson** [comment]
Ok. I'll wait for green, then will push to turn off verbose, and then you can presumably merge safely without waiting for the non-verbose builds to re-run.

**Comment by wesm** [comment]
SGTM. thanks!

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**
